### DISCUSSION TITLE
el-date-picker 不双向绑定怎么触发事件

### DISCUSSION THREAD

**Discussion by langziyang** [element-plus/element-plus#15132] [Q&A]
```
<el-date-picker
              clearable
              placeholder="结束时间"
              format="YYYY-MM-DD"
              :model-value="filter.before"
              value-format="YYYY-MM-DD"
              @change="handleChange"
              @calendar-change="handleChange"
          />
```
选择时间后没有哪一个事件被触发

**Comment by xing403** [comment]
我的理解是变量如果没有双向绑定，那么当变量发生变化时，Vue 不会检测到这个变量的变化，也就不会触发相应的 `change` 等事件。如果你希望在变量发生变化时触发 `change` 事件，你必须使用双向绑定监听变量的变化。

> 使用方法-->

```vue
<el-date-picker 
    v-model:model-value="filter.before" clearable placeholder="结束时间" format="YYYY-MM-DD"
    value-format="YYYY-MM-DD" @change="handleChange" @calendar-change="handleChange" 
/>
<!-- or -->
<el-date-picker 
    v-model="filter.before" clearable placeholder="结束时间" format="YYYY-MM-DD" value-format="YYYY-MM-DD"
    @change="handleChange" @calendar-change="handleChange" 
/>
```

**Reply by langziyang** [reply]
那如果我监听filter变化做别的请求，但是我又需要日期变化时传自定义的值到filter，不是没办法？

**Reply by xing403** [reply]
> 可以的，使用 `watch` 监听 `filter`
```vue
<script setup lang="ts">
import { ref, watch } from 'vue'
const filter = ref({
  before: null
})
// filter.value.before value is change
function handleChange(val: any) {
  console.log('handleChange', val)
}
// watch filter
watch(filter, (val: any) => {
  console.log('watch filter change', val)
}, { deep: true })
</script>

<template>
  <el-date-picker v-model:model-value="filter.before" clearable placeholder="结束时间" format="YYYY-MM-DD"
    value-format="YYYY-MM-DD" @change="handleChange" />
</template>
```
之后选择日期发生变化时: 结果--->
![image](https://github.com/element-plus/element-plus/assets/74746021/22c08741-53f9-41a9-b69a-b1c93c02d9db)

**Reply by langziyang** [reply]
对，我就是用watch监听filter,filter一改变就去做别的请求，但是我想说的是这个date-picker，如果绑定数据，一选择这个值，watch就监测到了，但是我需要在这个date-picker选择时，根据选择的值去做别的逻辑处理，然后再传别的值给filter

**Reply by xing403** [reply]
> 对，我就是用watch监听filter,filter一改变就去做别的请求，但是我想说的是这个date-picker，如果绑定数据，一选择这个值，watch就监测到了，但是我需要在这个date-picker选择时，根据选择的值去做别的逻辑处理，然后再传别的值给filter

那就是在`change`事件中处理before的改变，根据业务需求，使用把想要的值值赋予给一个新的变量，例如：`filter.age = 当前时间 - 选择时间`。个人不建议一个变量有两种不同的含义。

**Reply by langziyang** [reply]
实际上我想做的是当change时，不通过绑定传参数，因为我想通过我自己的方法比如说 updateFilter(key,value)，去更新filter里的数据。因为我在updateFilter里还有逻辑，似乎实现不了我这个要求，我不知道是element还是vue的原因

**Reply by xing403** [reply]
我觉得应该并没有什么影响，亦或者我未能正确理解你真正的业务逻辑需求。

**Reply by langziyang** [reply]
如果你在做列表时，有一个搜索和分页功能，就像我一样写一个公共方法，watch(filter,()=>{getList(filter)}), 当filter每一次改变，就去更新列表。现在的场景是：用户先选择分页到第三页，然后使用时间过滤，这个时候传给getList的page依然是第五页，但实际上应该是传page=1，所以如果我使用updateFilter(key,value)，我只需要判断key!==page,那我就可以手动修改filter.page=1，当然我知道我可以把page从filter分离出来，写不同的逻辑去处理，但是如果我仅通过updateFilter就能实现的话，我为什么还要去多写其它逻辑代码呢

**Reply by xing403** [reply]
我写了一个例子：依旧使用的watch,回调函数会给`新值`和`旧值`,请注意不要使用 `deep`因为 deep 的新值和旧值是相同的,就不能直接使用新值和旧值比较
官方说的，也确实表现的是这样的
![image](https://github.com/element-plus/element-plus/assets/74746021/ed534110-8396-4e28-991d-c6271d7c8550)
但如果直接监听 filter 会发现监听回调并未执行，(个人猜测是因为 fiter 是个对象，对象地址没有变，也就是没有触发这个回调)。可以通过监听具体的成员属性，来进行需要的操作
```vue
<script setup lang="ts">
import { ref, watch } from 'vue'
const filter = ref({
  date: null,
  page: 1,
  size: 10,
})
function handleChange(val: any) {
  console.log('handleChange', val)
}
// watch filter
watch(() => [
  filter.value.page,
  filter.value.size,
  filter.value.date
], (value: any, oldValue: any) => {
  // value and oldValue format: [page, size, date]
  console.log('value', value)
  console.log('oldValue', oldValue)
})
</script>

<template>
  <el-date-picker v-model="filter.date" clearable placeholder="日期" format="YYYY-MM-DD" value-format="YYYY-MM-DD"
    @change="handleChange" />
  <el-pagination v-model:current-page="filter.page" v-model:page-size="filter.size" layout="prev, pager, next"
    :total="100" />
</template>

```
改变前
![image](https://github.com/element-plus/element-plus/assets/74746021/6d133fdf-e61f-4d37-952c-d9925493fb01)
改变后
![image](https://github.com/element-plus/element-plus/assets/74746021/23cea65b-c726-4963-8a76-3532cf019ff2)
